prov/ucx/src/ucx_callbacks.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (c) 2016-2023 Intel Corporation. All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * OpenFabrics.org BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "ucx.h"
+#include "ucx_core.h"
+
+void ucx_send_callback_no_compl(void *request, ucs_status_t status)
+{
+	struct ucx_request *ucx_req = request;
+
+	ofi_ep_tx_cntr_inc(&ucx_req->ep->ep);
+	ucx_req_release(request);
+}
+
+void ucx_send_callback(void *request, ucs_status_t status)
+{
+	struct util_cq *cq;
+	struct ucx_request *ucx_req = request;
+	struct fi_cq_tagged_entry *tc = &ucx_req->completion.tagged;
+
+	cq = ucx_req->cq;
+	if (status != UCS_OK){
+		int ret;
+		UCX_PUSH_ERROR_COMPLETION(cq, tc->op_context, tc->flags,
+					  (int) status,
+					  -UCX_TRANSLATE_ERRCODE(status),
+					  0, ret, tc->tag);
+		(void) ret;
+	} else {
+		ofi_cq_write(cq, tc->op_context, tc->flags, tc->len,
+			     tc->buf, 0, tc->tag);
+		ofi_ep_tx_cntr_inc(&ucx_req->ep->ep);
+	}
+	ucx_req_release(request);
+}
+
+void ucx_recv_callback_no_compl(void *request, ucs_status_t status,
+				ucp_tag_recv_info_t *info)
+{
+	struct ucx_request *ucx_req = request;
+
+	ofi_ep_rx_cntr_inc(&ucx_req->ep->ep);
+	ucx_req_release(request);
+}
+
+void ucx_recv_callback(void *request, ucs_status_t status,
+		       ucp_tag_recv_info_t *info)
+{
+	struct util_cq *cq;
+	struct ucx_request *ucx_req;
+	struct fi_cq_tagged_entry *tc;
+
+	ucx_req = (struct ucx_request*)request;
+	ucx_req->completion.tagged.tag = info->sender_tag;
+	ucx_req->completion.tagged.len = info->length;
+	ucx_req->status = status;
+
+	if (ucx_req->type == UCX_FI_REQ_UNINITIALIZED) {
+		ucx_req->type = (status != UCS_OK) ?
```

Comment:
This looks like a place for an assert(I should never be here)

Reply:
Why? This got hit a lot in my test. This is basically the "unexpect message" case.

Reply:
Maybe there's a better choice than UNINITIALIZED then.

Reply:
How about UNKOWN? or UNSPEC?  It's set when the 'status' (request) object is allocated by ucp.